src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs

```diff
@@ -922,6 +1012,170 @@ public void WebApplicationBuilder_ThrowsFromExtensionMethodsNotSupportedByHostAn
+            }).WithDisplayName("One");
+
+            await app.StartAsync();
+
+            var ds = app.Services.GetRequiredService<EndpointDataSource>();
+            Assert.Equal(1, ds.Endpoints.Count);
+            Assert.Equal("One", ds.Endpoints[0].DisplayName);
+
+            var client = app.GetTestClient();
+
+            _ = await client.GetAsync("http://localhost/");
+            Assert.Equal("One", chosenRoute);
+        }
+
+        [Fact]
+        public async Task BranchingPipelineHasOwnRoutes()
+        {
+            var builder = WebApplication.CreateBuilder();
+            builder.WebHost.UseTestServer();
+            await using var app = builder.Build();
+
+            app.UseRouting();
+
+            var chosenRoute = string.Empty;
+            app.MapGet("/", () => "Hello World!").WithDisplayName("One");
+
+            app.UseEndpoints(routes =>
+            {
+                routes.MapGet("/hi", async c =>
+                {
+                    chosenRoute = c.GetEndpoint()?.DisplayName;
+                    await c.Response.WriteAsync("Hello World");
+                }).WithDisplayName("Two");
+                routes.MapGet("/heyo", () => "Heyo World").WithDisplayName("Three");
+            });
+
+            var newBuilder = ((IApplicationBuilder)app).New();
+            Assert.False(newBuilder.Properties.TryGetValue(WebApplication.GlobalEndpointRouteBuilderKey, out _));
+
+            newBuilder.UseRouting();
+            newBuilder.UseEndpoints(endpoints =>
+            {
+                endpoints.MapGet("/h3", async c =>
+                {
+                    chosenRoute = c.GetEndpoint()?.DisplayName;
+                    await c.Response.WriteAsync("Hello World");
+                }).WithDisplayName("Four");
+                endpoints.MapGet("hi", async c =>
+                {
+                    chosenRoute = c.GetEndpoint()?.DisplayName;
+                    await c.Response.WriteAsync("Hi New");
+                }).WithDisplayName("Five");
+            });
+            var branch = newBuilder.Build();
+            app.Run(c => branch(c));
+
+            app.Start();
+
+            var ds = app.Services.GetRequiredService<EndpointDataSource>();
+            Assert.Equal(5, ds.Endpoints.Count);
+            Assert.Equal("One", ds.Endpoints[0].DisplayName);
+            Assert.Equal("Two", ds.Endpoints[1].DisplayName);
+            Assert.Equal("Three", ds.Endpoints[2].DisplayName);
+            Assert.Equal("Four", ds.Endpoints[3].DisplayName);
+            Assert.Equal("Five", ds.Endpoints[4].DisplayName);
+
+            var client = app.GetTestClient();
+
+            // '/hi' routes don't conflict and the non-branched one is chosen
+            _ = await client.GetAsync("http://localhost/hi");
+            Assert.Equal("Two", chosenRoute);
+
+            // Can access branched routes
+            _ = await client.GetAsync("http://localhost/h3");
+            Assert.Equal("Four", chosenRoute);
+        }
+
```

Comment:
Can you add a test for the following scenario:

```C#

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddSingleton<IStartupFilter, Filter>();
var app = builder.Build();

((IApplicationBuilder)app).Properties["didsomething"] = true;

app.Run();

class Filter : IStartupFilter
{
    public Action<IApplicationBuilder> Configure(Action<IApplicationBuilder> next)
    {
        return app =>
        {
            next(app);

            // This should be true
            var val = app.Properties["didsomething"];
        };
    }
}
```

This verifies that an IStartupFilter can observe changes to the IApplicationBuilder (in properties).